File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
class Word{
    constructor(str, step){
        this.str=str;
        this.step=step;
    }
}

function solution(begin, target, words) {
    let totalWords = words.length;
    let isVisited = new Array(totalWords).fill(0);
    let q = [];
    q.push(new Word(begin, 0));
    
    function isNext(str1, str2){
        let unMatchCnt=0;
        for(let i=0; i<str1.length; i++){
            if(str1[i]!==str2[i]){
                unMatchCnt++;
                if(unMatchCnt>1) return false; 
            }
        }
        return true;
    }
    
    while(q.length!=0){
        let currentWord = q.shift();
        if(currentWord.str===target){
            return currentWord.step;
        }
        
        for(let i=0; i<totalWords; i++){
            if(!isVisited[i]&&isNext(currentWord.str, words[i])){
                isVisited[i]=1;
                q.push(new Word(words[i], 1+currentWord.step));
            }
        }
    }
    
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static class Word{
        String str;
        int step;
        
        Word(String str, int step){
            this.str=str;
            this.step=step;
        }
    }
    
    public boolean matchWords(String str1, String str2){
        int unMatch=0;
        for(int i=0; i<str1.length(); i++){
            if(str1.charAt(i)!=str2.charAt(i)){
                unMatch++;
                if(unMatch>=2) return false;
            }
        }
        return true;
    }
    
    public int solution(String begin, String target, String[] words) {
        int totalWords = words.length;
        boolean[] isVisited = new boolean[totalWords];
        
        Deque<Word> dq = new ArrayDeque<>();
        dq.offerLast(new Word(begin, 0));
        
        while(!dq.isEmpty()){
            Word currentWord = dq.pollFirst();
            if(currentWord.str.equals(target)){
                return currentWord.step;
            }
            for(int i=0; i<totalWords; i++){
                if(isVisited[i]) continue;
                String nextWord = words[i];
                if(matchWords(currentWord.str, nextWord)){
                    isVisited[i]=true;
                    dq.offerLast(new Word(nextWord, currentWord.step+1));
                }
            }
        }
        
        return 0;
    }
}